### ISSUE TITLE
Sets: support more than primitive types

### ISSUE THREAD

**Issue by pgte** [ipfs-shipyard/js-delta-crdts#15]
ORSets only support primitive JS types for uniqueness.
Devise some custom way to define equality / uniqueness.

/cc @parkan and @satazor

**Comment by satazor** [comment]
To support non-primitive types we must have a way to identity each item uniquely. We can do this implicitly by using tools like https://npms.io/search?q=hash-object but also opt-in for specifying the ids explicitly in the mutators argument.

Though, this will complicate the Sets CRDTs. For now, we should make it clear in the README that only primitive types are supported (for now).

**Comment by pgte** [comment]
Internally we use a DotSet, where the key is the object.
We can make it so that if the object has an `id` property, we use that as the instead.
Otherwise, the current behaviour happens.
@satazor makes sense?

**Comment by parkan** [comment]
in JVM-land this sort of thing is handled by defining a `.hashCode` method; this is nice because you can say things like "hash all value fields but not timestamp" -- sounds like we can have a `id` that's a hash of the object by default but can be overriden?

there's potentially concerns with things like field order so it'd be nice to have this handled robustly

**Comment by pgte** [comment]
@parkan that's a good analogy. I was going more towards defining distributed object identity, but I guess that what we want is value equivalence. I think that there's an opportunity to do that now in #17.

**Comment by satazor** [comment]
Regarding using `.hashCode`: I fear that the convention will force users to duplicate their "id" property that most of them will already have into `.hashCode`. What if we just create a option to compute the hash code, that by default uses `hash-it` but the user could customize it to something: `(item) => item.id`?

**Comment by pgte** [comment]
I've been giving this some thought and I see one problem:
These objects are going to be distributed and instantiated eventually without control from the user. 
For instance, a structure may have an ORMap, where one key can map to a AWORSet. When receiving this update, the replica does not know this in advance, and has not opportunity to define this.

If the user could define this customisation, lack of different configuration in replicas will lead to divergence (ouch!).

Thoughts?

**Comment by satazor** [comment]
> For instance, a structure may have an ORMap, where one key can map to a AWORSet. When receiving this update, the replica does not know this in advance, and has not opportunity to define this.

Could you elaborate this point?

> If the user could define this customisation, lack of different configuration in replicas will lead to divergence (ouch!).

Indeed, but this is solvable by having the "hash code" be computed at insertion time instead of being calculated by each replicate. This is similar to having the `.hashCode` property but it's hidden from the outside.

**Comment by pgte** [comment]
@satazor, yes, but it's a convention. The problem I'm describing can be described as something like:

I'm replica A. I define a custom hash function. I add an element to a set. I replicate that element to replica B.
When replica B receives this value, it doesn't know which hash function to use, thus losing coherency.

Makes sense?

**Comment by satazor** [comment]
@pgte I understand, but how is computing the hashcode by the replica adding it different from having the hashCode property set manually? The other replicas won’t be able to check if it’s correct because they don’t have the hashing function too, right?

**Comment by pgte** [comment]
To clarify:
It's not a problem of checking, it's a problem of the uniqueness being a function that is defined outside of the CRDT itself.

**Comment by satazor** [comment]
> To clarify:
> It's not a problem of checking, it's a problem of the uniqueness being a function that is defined outside of the CRDT itself.

Both strategies, `hashCode` via a property or an option, live outside the scope of the CRDT. That's what I've been trying to say:

```js
replica.add({
  id: 'abc', 
  foo: 'bar',
  hashCode: 'abc',
})

// vs

const replicate = CRDT('my replica', {
   hashCode: (item) => item.id,  // Default being hashIt(item) for objects
});

replica.add({
  id: 'abc', 
  foo: 'bar',
})
```

In both scenarios, the  hashCode is being set explicitly and both share the same problem. What if the replicas are running different hashing functions?

While they share the same problem, by having the option we are giving`delta-crdts` the opportunity to calculate the hashCode of items being merged by other replicas. This would allow to emit a warning when that happens, at least.
